csharp/src/Ice/Communicator.cs

```diff
@@ -896,8 +911,27 @@ async Task PerformDisposeAsync()
 
                 // Shutdown and destroy all the incoming and outgoing Ice connections and wait for the connections to be
                 // finished.
-                await Task.WhenAll(OutgoingConnectionFactory?.DisposeAsync().AsTask() ?? Task.CompletedTask,
-                                   ShutdownAsync()).ConfigureAwait(false);
+                IEnumerable<Task> tasks =
+                    _outgoingConnections.Values.SelectMany(connections => connections).Select(connection =>
+                        connection.GoAwayAsync(new CommunicatorDisposedException())).Append(
+                        ShutdownAsync());
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+
+#if DEBUG
```

Comment:
Should we link the communicator token to the ConnectTimeout token, so that all pending connections are canceled upon calling Dispose of the communicator?

Reply:
Yes, that's simpler!